cmd/proxy.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/cloudsqlconn"
+	"github.com/spf13/cobra"
+)
+
+// proxyClient represents the state of the current instantiation of the proxy.
+type proxyClient struct {
+	cmd    *cobra.Command
+	dialer *cloudsqlconn.Dialer
+
+	// mnts is a list of all mounted sockets for this client
+	mnts []*socketMount
+	// serveCtx is used for serving data
+	serveCtx    context.Context
+	serveCancel context.CancelFunc
+}
+
+// newProxyClient completes the initial setup required to get the proxy to a "steady" state.
+func newProxyClient(ctx context.Context, cmd *cobra.Command, args []string) (*proxyClient, error) {
+	dialer, err := cloudsqlconn.NewDialer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing dialer: %v", err)
+	}
+	pc := &proxyClient{cmd: cmd, dialer: dialer}
+
+	port := 5000 // TODO: figure out better port allocation strategy
+	for i, inst := range args {
+		m := newSocketMount(dialer, inst)
+		addr, err := m.Listen(ctx, "tcp4", net.JoinHostPort("", fmt.Sprint(port+i)))
+		if err != nil {
+			return nil, fmt.Errorf("[%s] Unable to mount socket: %v", inst, err)
+		}
+		cmd.Printf("[%s] Listening on %s\n", inst, addr.String())
+		pc.mnts = append(pc.mnts, m)
+	}
+
+	return pc, nil
+}
+
+// Serve listens on the mounted ports and beging proxying the connections to the instances.
+func (pc *proxyClient) Serve(ctx context.Context) error {
+	pc.serveCtx, pc.serveCancel = context.WithCancel(ctx)
```

Comment:
I think we can avoid saving a reference to cancel and the context here, by moving this cancelation logic to where Serve is called below.

``` golang
go func() {
  ctx2, cancel := context.WithCancelation(ctx)
  defer cancel()
  shutdownCh <- p.Serve(ctx2)
}
```

Reply:
I don't think we want to do that because at some point "Close" is going to do a bunch of work to clean up before forcing the connections to stop.

Reply:
Let's see where this goes and revisit. There's a post on this subject: https://go.dev/blog/context-and-structs. Let's see if that applies to our case or not in future PRs.